### PR TITLE
Move Cache-Control header setting before WriteHeader calls

### PR DESCRIPTION
**Pull request by rolandshoemaker**
#531 wrote the headers after `WriteHeader` calls, which in normal operation mean they won't be properly set and returned to the client. Since `httptest` seems to not properly follow this behavior tests still passed but the headers would [never actually be returned](https://godoc.org/net/http#ResponseWriter).

This moves the adding of the headers before any `WriteHeader` or `Write` calls so that the they are properly added to the response. Sorry about that!

### CODE REVIEW

**Comment by lziest** [comment]
Thanks, rule of thumb for me these days is never review code after 8pm. LGTM

**Comment by lziest** [comment]
@riking please review too.

**Comment by riking** [comment]
Why not use Header.Set() at the top and overwrite it once you know the max-age value?

**Comment by riking** [comment]
so:

```
response.Header().Set("Cache-Control", "max-age=0, no-cache")
// ...
if now.Before(parsedResponse.NextUpdate) {
	expiresIn := int(parsedResponse.NextUpdate.Sub(now) / time.Second)
	response.Header().Set("Cache-Control", fmt.Sprintf(..., expiresIn))
} else {
	// XXX: we want max-age=0 but since this is technically an authorized OCSP response
	// (despite being stale) and 5019 forbids attaching no-cache we get a little tricky
	response.Header().Set("Cache-Control", "max-age=0, public, no-transform, must-revalidate")
}
```

File: ocsp/responder.go [hunk 1]
```diff
@@ -195,10 +202,10 @@ func (rs Responder) ServeHTTP(response http.ResponseWriter, request *http.Reques
 	now := rs.clk.Now()
 	if now.Before(parsedResponse.NextUpdate) {
 		expiresIn := int(parsedResponse.NextUpdate.Sub(now) / time.Second)
-		maxAge = &expiresIn
+		addCacheHeader(response, &expiresIn)
 	} else {
-		zero := 0      // XXX: we want max-age=0 but since this is technically an authorized OCSP response
-		maxAge = &zero //      (despite being stale) and 5019 forbids attaching no-cache we get a little tricky
+		zero := 0                       // XXX: we want max-age=0 but since this is technically an authorized OCSP response
```
Comment (jsha) [inline ocsp/responder.go]: At least in Boulder, we use XXX to indicate "Don't merge this," and TODO to indicates long-term TODO's in merged code.

File: ocsp/responder.go [hunk 2]
```diff
@@ -93,6 +93,17 @@ func NewResponder(source Source) *Responder {
 	}
 }
 
+func addCacheHeader(response http.ResponseWriter, maxAge *int) {
```
Comment (jsha) [inline ocsp/responder.go]: This pattern of passing a pointer to an int is weird. How about instead take a regular int and:

```
if maxAge == 0 {
  ...max-age=0, no-cache
} else {
  ...max-age=%d
```

**Comment by rolandshoemaker** [comment]
Pushed simplified logic.

**Comment by jsha** [comment]
LGTM

**Comment by lziest** [comment]
lgtm, can you squash it?

**Comment by rolandshoemaker** [comment]
Done.

**Comment by jsha** [comment]
I missed a comment on this: Since `httptest` isn't capable of testing the bug here, is there another way of writing the test such that the bug actually does get caught?

**Pull request merged**